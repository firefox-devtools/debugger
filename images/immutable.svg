<!-- This Source Code Form is subject to the terms of the Mozilla Public
   - License, v. 2.0. If a copy of the MPL was not distributed with this
   - file, You can obtain one at http://mozilla.org/MPL/2.0/. -->
<!DOCTYPE svg PUBLIC "-//W3C//DTD SVG 20010904//EN"
 "http://www.w3.org/TR/2001/REC-SVG-20010904/DTD/svg10.dtd">
<svg version="1.0" xmlns="http://www.w3.org/2000/svg"
 width="200.000000pt" height="200.000000pt" viewBox="0 0 200.000000 200.000000"
 preserveAspectRatio="xMidYMid meet">

<g transform="translate(0.000000,200.000000) scale(0.100000,-0.100000)"
fill="#000000" stroke="none">
<path d="M247 1663 c-4 -3 -7 -235 -7 -515 l0 -508 175 0 175 0 0 515 0 515
-168 0 c-93 0 -172 -3 -175 -7z m183 -508 c0 -336 -2 -385 -15 -385 -13 0 -15
49 -15 385 0 336 2 385 15 385 13 0 15 -49 15 -385z"/>
<path d="M697 1664 c-4 -4 -7 -236 -7 -516 l0 -508 165 0 165 0 2 181 3 181
99 -146 c54 -80 102 -146 106 -146 3 0 51 66 106 148 l99 147 3 -182 2 -183
160 0 160 0 0 515 0 515 -132 0 -133 -1 -130 -194 c-71 -107 -132 -194 -135
-194 -3 -1 -64 86 -135 192 l-131 192 -130 3 c-71 1 -133 -1 -137 -4z m356
-374 c92 -138 172 -250 176 -250 4 0 84 112 177 250 112 166 175 250 187 250
16 0 17 -25 17 -385 0 -336 -2 -385 -15 -385 -13 0 -15 46 -15 350 0 193 -4
350 -8 350 -4 0 -82 -108 -172 -240 -90 -132 -167 -240 -171 -240 -4 0 -81
108 -171 240 -90 131 -166 239 -170 240 -5 0 -8 -157 -8 -350 0 -304 -2 -350
-15 -350 -13 0 -15 49 -15 385 0 360 1 385 18 385 11 -1 74 -86 185 -250z"/>
</g>
</svg>
