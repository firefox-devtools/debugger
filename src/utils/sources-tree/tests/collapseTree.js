import { Map } from "immutable";

import { collapseTree, formatTree, addToTree, createNode } from "../index";

const abcSource = Map({
  url: "http://example.com/a/b/c.js",
  actor: "actor1"
});
const abcdeSource = Map({
  url: "http://example.com/a/b/c/d/e.js",
  actor: "actor2"
});
const abxSource = Map({
  url: "http://example.com/a/b/x.js",
  actor: "actor3"
});

describe("sources tree", () => {
  describe("collapseTree", () => {
    it("can collapse a single source", () => {
      const fullTree = createNode("root", "", []);
      addToTree(fullTree, abcSource);
      expect(fullTree.contents.length).toBe(1);
      const tree = collapseTree(fullTree);

      const host = tree.contents[0];
      expect(host.name).toBe("example.com");
      expect(host.contents.length).toBe(1);

      const abFolder = host.contents[0];
      expect(abFolder.name).toBe("a/b");
      expect(abFolder.contents.length).toBe(1);

      const abcNode = abFolder.contents[0];
      expect(abcNode.name).toBe("c.js");
      expect(abcNode.path).toBe("/example.com/a/b/c.js");
      expect(formatTree(tree)).toMatchSnapshot();
    });

    it("correctly merges in a collapsed source with a deeper level", () => {
      const fullTree = createNode("root", "", []);
      addToTree(fullTree, abcSource);
      addToTree(fullTree, abcdeSource);
      const tree = collapseTree(fullTree);

      expect(tree.contents.length).toBe(1);

      const host = tree.contents[0];
      expect(host.name).toBe("example.com");
      expect(host.contents.length).toBe(1);

      const abFolder = host.contents[0];
      expect(abFolder.name).toBe("a/b");
      expect(abFolder.contents.length).toBe(2);

      const [cdFolder, abcNode] = abFolder.contents;
      expect(abcNode.name).toBe("c.js");
      expect(abcNode.path).toBe("/example.com/a/b/c.js");
      expect(cdFolder.name).toBe("c/d");

      const [abcdeNode] = cdFolder.contents;
      expect(abcdeNode.name).toBe("e.js");
      expect(abcdeNode.path).toBe("/example.com/a/b/c/d/e.js");
      expect(formatTree(tree)).toMatchSnapshot();
    });

    it("correctly merges in a collapsed source with a shallower level", () => {
      const fullTree = createNode("root", "", []);
      addToTree(fullTree, abcSource);
      addToTree(fullTree, abxSource);
      const tree = collapseTree(fullTree);

      expect(tree.contents.length).toBe(1);

      const host = tree.contents[0];
      expect(host.name).toBe("example.com");
      expect(host.contents.length).toBe(1);

      const abFolder = host.contents[0];
      expect(abFolder.name).toBe("a/b");
      expect(abFolder.contents.length).toBe(2);

      const [abcNode, abxNode] = abFolder.contents;
      expect(abcNode.name).toBe("c.js");
      expect(abcNode.path).toBe("/example.com/a/b/c.js");
      expect(abxNode.name).toBe("x.js");
      expect(abxNode.path).toBe("/example.com/a/b/x.js");
      expect(formatTree(tree)).toMatchSnapshot();
    });

    it("correctly merges in a collapsed source with the same level", () => {
      const fullTree = createNode("root", "", []);
      addToTree(fullTree, abcdeSource);
      addToTree(fullTree, abcSource);
      const tree = collapseTree(fullTree);

      expect(tree.contents.length).toBe(1);

      const host = tree.contents[0];
      expect(host.name).toBe("example.com");
      expect(host.contents.length).toBe(1);

      const abFolder = host.contents[0];
      expect(abFolder.name).toBe("a/b");
      expect(abFolder.contents.length).toBe(2);

      const [cdFolder, abcNode] = abFolder.contents;
      expect(abcNode.name).toBe("c.js");
      expect(abcNode.path).toBe("/example.com/a/b/c.js");
      expect(cdFolder.name).toBe("c/d");

      const [abcdeNode] = cdFolder.contents;
      expect(abcdeNode.name).toBe("e.js");
      expect(abcdeNode.path).toBe("/example.com/a/b/c/d/e.js");
      expect(formatTree(tree)).toMatchSnapshot();
    });
  });
});
