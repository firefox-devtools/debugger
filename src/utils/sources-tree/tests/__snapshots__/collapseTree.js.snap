// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sources tree collapseTree can collapse a single source 1`] = `
" - root path= 
  - example.com path=/example.com 
    - a/b path=/example.com/a/b 
      - c.js path=/example.com/a/b/c.js source_id=undefined 
"
`;

exports[`sources tree collapseTree correctly merges in a collapsed source with a deeper level 1`] = `
" - root path= 
  - example.com path=/example.com 
    - a/b path=/example.com/a/b 
      - c/d path=/example.com/a/b/c/d 
        - e.js path=/example.com/a/b/c/d/e.js source_id=undefined 
      - c.js path=/example.com/a/b/c.js source_id=undefined 
"
`;

exports[`sources tree collapseTree correctly merges in a collapsed source with a shallower level 1`] = `
" - root path= 
  - example.com path=/example.com 
    - a/b path=/example.com/a/b 
      - c.js path=/example.com/a/b/c.js source_id=undefined 
      - x.js path=/example.com/a/b/x.js source_id=undefined 
"
`;

exports[`sources tree collapseTree correctly merges in a collapsed source with the same level 1`] = `
" - root path= 
  - example.com path=/example.com 
    - a/b path=/example.com/a/b 
      - c/d path=/example.com/a/b/c/d 
        - e.js path=/example.com/a/b/c/d/e.js source_id=undefined 
      - c.js path=/example.com/a/b/c.js source_id=undefined 
"
`;
